lib/src/component2_suggestors/defaultprops_initialstate_migrator.dart

```diff
@@ -0,0 +1,168 @@
+        var methodBody = (node.body as BlockFunctionBody).block;
+
+        if (methodBody.statements.length == 1 &&
+            methodBody.statements.single is ReturnStatement) {
+          var returnStatement =
+              (methodBody.statements.single as ReturnStatement);
+
+          updateSuperCalls(returnStatement.returnKeyword.end,
+              returnStatement.semicolon.offset);
+
+          // Convert to arrow function if method body is a single return.
+          yieldPatch(
+            methodBody.leftBracket.offset,
+            returnStatement.returnKeyword.end,
+            '=> (',
+          );
+          yieldPatch(
+            returnStatement.semicolon.offset,
+            returnStatement.semicolon.offset,
+            '\n)',
+          );
+          yieldPatch(
+            methodBody.rightBracket.offset,
+            methodBody.rightBracket.end,
+            '',
+          );
+        } else {
+          updateSuperCalls(
+              methodBody.leftBracket.end, methodBody.rightBracket.offset);
+        }
+      } else if (node.body is ExpressionFunctionBody) {
+        var expression = (node.body as ExpressionFunctionBody).expression;
+
+        updateSuperCalls(expression.offset, expression.end);
+
+        String subMethod;
+        if (methodToMigrate == 'getDefaultProps') {
+          subMethod = 'newProps';
+        } else {
+          subMethod = 'newState';
+        }
+
+        // Add parenthesis if needed.
+        if (expression is! ParenthesizedExpression &&
+            expression is CascadeExpression &&
+            expression.target is MethodInvocation &&
+            (expression.target as MethodInvocation).methodName.name ==
+                subMethod) {
+          yieldPatch(
+            expression.offset,
+            expression.offset,
+            '(',
+          );
+          yieldPatch(
+            expression.end,
+            expression.end,
+            '\n)',
+          );
+        }
+      }
+    }
+  }
+
+  /// Updates all super calls to [methodToMigrate] from [start] to [end].
+  void updateSuperCalls(int start, int end) {
+    var methodBodyString = sourceFile.getText(start, end);
+    if (methodBodyString.contains('super.$methodToMigrate()')) {
+      methodBodyString = methodBodyString.replaceAll(
+        'super.$methodToMigrate()',
+        'super.$migrateTo',
+      );
+      yieldPatch(start, end, methodBodyString);
+    }
+  }
+}
+
+/// Suggestor that replaces `getDefaultProps` method with getter `defaultProps`.
+class GetDefaultPropsMigrator extends DefaultPropsInitialStateMigrator
+    implements Suggestor {
+  GetDefaultPropsMigrator({
+    allowPartialUpgrades = true,
+    shouldUpgradeAbstractComponents = false,
+  }) {
+    this.allowPartialUpgrades = allowPartialUpgrades;
+    this.shouldUpgradeAbstractComponents = shouldUpgradeAbstractComponents;
+    this.methodToMigrate = 'getDefaultProps';
+    this.migrateTo = 'defaultProps';
+  }
```

Comment:
Instead of using non-final fields in the abstract class, I would prefer that we make them final, and utilize a constructor in the abstract like so:

```dart
abstract class DefaultPropsInitialStateMigrator extends GeneralizingAstVisitor
    with AstVisitingSuggestorMixin
    implements Suggestor {
  final bool allowPartialUpgrades;
  final bool shouldUpgradeAbstractComponents;
  final String methodToMigrate;
  final String migrateTo;

  DefaultPropsInitialStateMigrator({
    @required this.methodToMigrate,
    @required this.migrateTo,
    this.allowPartialUpgrades = true,
    this.shouldUpgradeAbstractComponents = false,
  });
}

/// Suggestor that replaces `getDefaultProps` method with getter `defaultProps`.
class GetDefaultPropsMigrator extends DefaultPropsInitialStateMigrator
    implements Suggestor {
  GetDefaultPropsMigrator({
    bool allowPartialUpgrades = true,
    bool shouldUpgradeAbstractComponents = false,
  }) : super(
    methodToMigrate: 'getDefaultProps', 
    migrateTo: 'defaultProps', 
    allowPartialUpgrades: allowPartialUpgrades,
    shouldUpgradeAbstractComponents: shouldUpgradeAbstractComponents,
  );
}

/// Suggestor that replaces `getInitialState` method with getter `initialState`.
class GetInitialStateMigrator extends DefaultPropsInitialStateMigrator
    implements Suggestor {
  GetInitialStateMigrator({
    bool allowPartialUpgrades = true,
    bool shouldUpgradeAbstractComponents = false,
  }) : super(
    methodToMigrate: 'getInitialState', 
    migrateTo: 'initialState', 
    allowPartialUpgrades: allowPartialUpgrades,
    shouldUpgradeAbstractComponents: shouldUpgradeAbstractComponents,
  );
}
```

Reply:
That makes a lot more sense!